### PR TITLE
Tika 2909

### PR DESCRIPTION
**Pull request by soomyung**
I want to contribute HWP v5 Parser

### CODE REVIEW

File: tika-parsers/pom.xml [hunk 1]
```diff
@@ -978,6 +978,19 @@
                     <ignore />
                   </action>
                 </pluginExecution>
+                <pluginExecution>
```
Comment (tballison) [inline tika-parsers/pom.xml]: Is this necessary?

File: tika-parsers/pom.xml [hunk 2]
```diff
@@ -969,7 +969,7 @@
                   <pluginExecutionFilter>
                     <groupId>org.codehaus.gmaven</groupId>
                     <artifactId>groovy-maven-plugin</artifactId>
-                    <version>${groovy.maven.version}</version>
+                    <versionRange>${groovy.maven.version}</versionRange>
```
Comment (tballison) [inline tika-parsers/pom.xml]: Why do we need this?

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpStreamReader.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	/**
+	 * 
+	 * @param n
+	 * @return
+	 * @throws IOException
+	 */
+	public long skip(long n) throws IOException {
+		return input.skip(n);
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpStreamReader.java]: Please use IOUtils.skipFully() ...

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpV5Parser.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+	private static final long serialVersionUID = 1L;
+	
+	private static final Set<MediaType> SUPPORTED_TYPES = Collections.singleton(MediaType.application("x-hwp-v5"));
+    public static final String HWP_MIME_TYPE = "application/x-hwp-v5";
+    
+    private HwpTextExtractorV5 extractor;
+    
+    public HwpV5Parser() {
+    	extractor = new HwpTextExtractorV5();
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpV5Parser.java]: Parsers need to be thread-safe.  HwpTextExtractorV5 stores a poifs as an instance variable.  So, you need to create a new HwpTextExtractorV5 for each parse, or remove the poifs instance variable.

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/UnsupportedHwpFormatException.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.tika.parser.hwp;
+
+/**
+ * when cannot parse HWP format.
+ */
+public class UnsupportedHwpFormatException extends Exception {
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/UnsupportedHwpFormatException.java]: Can you make this extend a TikaException?

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java [hunk 6]
```diff
@@ -0,0 +1,520 @@
...
+		} else {
+			parseBodyText(header, root, xhtml);
+		}
+	}
+
+	private void parseSummaryInformation(DirectoryNode root, Metadata metadata) throws UnsupportedHwpFormatException {
+
+		try {
+			
+			Entry summaryEntry = root.getEntry("HwpSummaryInformation");
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: Can you modify this to "\u0000HwpSummaryInformation" or whatever the correct unicode code is for the first character?

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java [hunk 7]
```diff
@@ -0,0 +1,520 @@
...
+				(DocumentEntry) summaryEntry);
+		
+		PropertySet ps = new PropertySet(summaryStream);
+		
+		Property[] props = ps.getProperties();
+		
+		for(Property prop : props) {
+			int propID = (int)prop.getID();
+			Object value = prop.getValue();
+
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: Check for null value?

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java [hunk 8]
```diff
@@ -0,0 +1,520 @@
...
+					&& entry instanceof DocumentEntry) {
+				log.debug("extract {}", entry.getName());
+
+				InputStream input = new DocumentInputStream(
+						(DocumentEntry) entry);
+				if (header.compressed)
+					input = new InflaterInputStream(input, new Inflater(true));
+
+				HwpStreamReader reader = new HwpStreamReader(input);
+
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: Add xhtml markup <div class="section"> ?

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java [hunk 9]
```diff
@@ -0,0 +1,520 @@
...
+			} else {
+				log.warn("unknown Entry '{}'({})", entry.getName(), entry);
+			}
+		}
+	}
+
+	private Key readKey(InputStream input) throws IOException {
+		byte[] data = new byte[260];
+
+		input.read(data, 0, 4); // TAG,
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: Use IOUtils readFully and check for read length if readfully doesn't.

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java [hunk 10]
```diff
@@ -0,0 +1,520 @@
...
+			if (n == 0) {
+				xor = (byte) (srand.rand() & 0xFF);
+				n = (int) ((srand.rand() & 0xF) + 1);
+			}
+			if (i >= 4) {
+				data[i] = (byte) ((data[i]) ^ (xor));
+			}
+		}
+
+		int offset = 4 + (data[0] & 0xF);
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: add sanity/range check to offset

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java [hunk 11]
```diff
@@ -0,0 +1,520 @@
...
+				writeParaHeader(reader, tag.length, buf);
+			} else if (HWPTAG_BEGIN + 51 == tag.id) {
+				if (tag.length % 2 != 0)
+					throw new IOException("Invalid block size");
+
+				writeParaText(reader, tag.length, buf);
+
+				if (buf.length() > 0) {
+					buf.append('\n');
+					xhtml.characters(buf.toString());
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: Add xhtml p element <p>

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java [hunk 12]
```diff
@@ -0,0 +1,520 @@
...
+	 * transfer character stream of HWPTAG_PARA_TEXT to STRING
+	 * 
+	 * @param reader
+	 * @param datasize
+	 * @param buf
+	 * @throws IOException
+	 */
+	private void writeParaText(HwpStreamReader reader, long datasize,
+			StringBuffer buf) throws IOException {
+		int[] chars = reader.uint16((int) (datasize / 2));
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: Somewhere do a sanity/range check on datasize...apologies if you've done that earlier!

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpStreamReader.java [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+	/**
+	 * count만큼 바이트를 읽는다. InflaterInputStream의 경우 한번에 count만큼 read가 안되는 경우가 있다.
+	 * 그래서 count만큼 읽을 때까지 루프를 실행한다
+	 * 
+	 * @param count
+	 * @return
+	 * @throws IOException
+	 * @throws EOFException
+	 */
+	private int ensure(int count) throws IOException, EOFException {
```
Comment (tballison) [inline tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpStreamReader.java]: Perhaps us IOUtils.readFully?  It is great to see that you are doing this!

File: tika-parsers/pom.xml (hunk 1 above)
Comment (soomyung) [reply tika-parsers/pom.xml]: This is not necessary. I didn't mean to

File: tika-parsers/pom.xml (hunk 2 above)
Comment (soomyung) [reply tika-parsers/pom.xml]: Eclipse failed to build workspace with "version".

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/UnsupportedHwpFormatException.java (hunk 5 above)
Comment (soomyung) [reply tika-parsers/src/main/java/org/apache/tika/parser/hwp/UnsupportedHwpFormatException.java]: Removed this class by ddoleye

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpV5Parser.java (hunk 4 above)
Comment (soomyung) [reply tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpV5Parser.java]: removed the poifs instance variable

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java (hunk 12 above)
Comment (soomyung) [reply tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: The reader.uint16() has a sanity/range check.

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpStreamReader.java (hunk 13 above)
Comment (soomyung) [reply tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpStreamReader.java]: modified as your direction

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java (hunk 11 above)
Comment (soomyung) [reply tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: modified as your direction

File: tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java (hunk 10 above)
Comment (soomyung) [reply tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: I think sanity/range check to offset is not necessary.  data[] has the fixed length of 260 and offset has the range of 4 ~ 19

Comment (tballison) [reply tika-parsers/src/main/java/org/apache/tika/parser/hwp/HwpTextExtractorV5.java]: doh...sorry....thank you!

**Pull request merged**